src/hotspot/share/opto/addnode.cpp

```diff
@@ -769,6 +771,553 @@ uint AddPNode::match_edge(uint idx) const {
   return idx > Base;
 }
 
+/* MergeLoads optimization
+ * Optimize multiple loads into a merged load, like below java code:
+ *
+ *      return ((array[offset    ] & 0xff) << 24)
+ *           | ((array[offset + 1] & 0xff) << 16)
+ *           | ((array[offset + 2] & 0xff) <<  8)
+ *           | ((array[offset + 3] & 0xff)      );
+ *
+ * The C2 IR graph is like:
+ *
+ *       Mem
+ *        |---------+---------+----------+
+ *        |         |         |          |
+ *      LoadB1    LoadB2    LoadB3     LoadB4
+ *        |         |         |          |
+ *        |       LShift    LShift     LShift
+ *        |         |         |          |
+ *        +-> Or1 <-+         |          |
+ *             |              |          |
+ *             +---> Or2 <----+          |
+ *                    |                  |
+ *                    +-----> Or3 <------+
+ *
+ * It is transformed as a merged LoadI, which replace the Or3 node.
+ *
+ * Because array value is logical and with constant 0xff/0xffff, LoadS/LoadB is converted to an unsigned load
+ * and 'And' node is eliminated in previous IGVN phase. Please check AndINode::Ideal for reference
```

Comment:
What do you mean by `array value is logical`?
Do you have tests where we do not have the `And` in the code, and there would be implicit sign extension?

Reply:
I mean `array value is masked with constant 0xff/0xffff`. I will change the comment. Tests without `And operator` will be added. I think the sign extension is checked before merging.